### ISSUE TITLE
Design and implement a dynamic membrane rendering around microbes

### ISSUE THREAD

**Issue by jjonj** [Revolutionary-Games/Thrive#115]
This will be very difficult and will require quite some experience with ogre/graphics I imagine.
The way to do it will probably be vertex buffer modification through ogre based on the hexes in the microbes compound scene node.

**Comment by jjonj** [comment]
Had any luck patowen?

**Comment by patowen** [comment]
I've managed to set it up, but I haven't worked on it. I seem to lose steam as soon as I go through the ordeal of trying to set it up. I'm also not sure whether the current algorithm I have in place is good enough, since it is not as efficient as I would like, and the resulting shape cannot easily be adjusted by changing parameters. A possible algorithm was given in the forums, but it just creates a 2D path, and as far as I know, there is no generic algorithm for converting that to a 3D mesh (besides making a cylinder or something like that).

The problem with metaballs is that unless the equation for them is very finely tuned, a straight line of hexes will create a bumpy cell surface. Perhaps when I have the time and energy, I may experiment with something else. I'll let you know if I find anything good.

**Comment by jjonj** [comment]
Sounds fair enough! Efficiency isn't too big of a deal when doing stuff in the microbe editor and regarding bumpy surfaces then we talked about not having to align the cytoplasm in hexes, I'd imagine that this would help!

**Comment by jjonj** [comment]
tjwhale created a post about a proof of concept for cell deformation: http://thrivegame.forum-free.ca/t1508-cell-membrane#31501 that perhaps could be incorporated! I'm not really sufficiently experienced in the subject to judge for myself at this point.

**Comment by Moopli** [comment]
Thing is, we don't *need* to finely tune the metaball surface -- we can convert the metaball funcion to an SDF, add some smooth noise, and the surface generated by marching-cubes will have much less regular bumpiness. Then, we can do a bunch of vertex post-processing, smoothing/roughening the surface further.

Right now, we absolutely don't need the surface produced by the metaballs to look perfectly-tuned. We just need it fully hooked-up and rendering in-game.

As for solutions that don't build the skin from a volumetric function (for example, the drawing-splines stuff), I have yet to see any solution that will give us a complete, smooth surface with the amount of customized shape we need. 

tjwhale's idea is pretty much generating coherent noise over a sphere -- it looks nice with one control point (the center of the circle), but add more control points and you pretty much have to define a metaball function anyway to get a smooth "average surface" over which you apply the noise. In the end it's a suboptimally-efficient way to noise a mesh. We still have to get the mesh somehow.

**Comment by Moopli** [comment]
Right after I wrote that comment I came up with an idea: Say you draw your microbe's skin in profile (ie, a closed, self-intersection-free curve that defines the intersection of the microbe surface with the plane which results in the largest enclosed area). Find the SDF of this curve, and you get a good starting measure of how fat your microbe should be on that part of the cross-section.  Obviously, that has to be processed further to give us a depth, to remove the sharp edges and make it all nice and bulbous. This would just be different way to produce the volume that we run marching cubes on, of course, so all the post-processing mesh-stretching stuff can still be done.

**Comment by patowen** [comment]
That might work. However, I still think that ideally, we could bypass marching cubes all together and start out with a mesh. It could be a simple, almost cylindrical mesh that we then subdivide and smooth.

**Comment by Moopli** [comment]
Well, one easy way to speed it up would be to generate at much lower resolution, and then project the generated vertices to the implicit surface. We could get, probably, at least an order of magnitude reduction in mesh size that way, since the surface shouldn't need any areas of fidelity/curvature high enough to justify the current sampling resolution.

**Comment by DrChewbacca** [comment]
Hi guys, TJWhale here, I've updated my thread to show an example with multiple control points. Hope you like it!

**Comment by Moopli** [comment]
@patowen I've been thinking about bypassing marching cubes, and here's what I've come up with:

I think it would be difficult and annoying to generate a mesh by subdivision if we stick to a surface defined by hex-grid metaballs. Luckily, it should be possible if instead of metaballs we use some sort of cross-section spline thing like I described above.

Once we've generated a nice, blobby volume in SDF form, we can generate a starting mesh by taking a spherical mesh and projecting each vertex outwards from the center to the actual surface. In areas of sufficient roughness, we subdivide and project recursively. We generate an SDF because projection is very easy when using one -- you read the projection distance directly off the SDF, and calculate the gradient to get the direction.

I'm not sure if there's an easy, efficient way to calculate the SDF of an arbitrary solid -- if not, then we'd project points to the surface using some on-the-fly binary searching.

**Comment by DrChewbacca** [comment]
Sounds a little like what I've been doing in my thread. Not sure if I'm on the right track at all. Have a look at the 3d stuff.

http://thrivegame.forum-free.ca/t1508-cell-membrane#31501

**Comment by Moopli** [comment]
I haven't had the time to go through your code @DrChewbacca, so I can't really judge (also the reason I haven't commented). How are you generating new surface points? So far it looks nice, but the important feature missing is the triangulation of the points. Depending on how you're generating new points, this might be easy or very hard.

**Comment by DrChewbacca** [comment]
Not sure what "triangulation of points" means, sorry. What I've been doing is generating the net using polar coordinates with radius and angle(s).

The algorithm is

1) Generate a net, evenly spaced by angle(s), at random radius not less than the distance from the origin to the furthest organelle.
2) Smooth the net*
3) If any point too close to an organelle (euclidean distance in space) then it is moved further away from the origin (it's radius is increased).
4) Possibly shrink wrap a bit (move all points closer to the origin).
goto 2 

*the way the smoothing works is just via the heat equation but the best way to think of it is "each point tries to be the average distance from the origin of it's neighbours".

So for any point it knows it's radius (r) and it's neighbours radii (r_n for 1..n) and then calculates

change in r = 0.1* (sum (r_n,n) - n*r)

so if there are no organelles then the smoothing will make the net into a sphere but if there are any then (3) will mean there are lumps. But it will always be smooth because any local noise or spikes are smoothed out very quickly.

Hope this makes sense, I'll happily explain again if this is hard to understand.

**Comment by Moopli** [comment]
An example of the sort of triangulation I mean is the [Delaunay Triangulation](http://en.wikipedia.org/wiki/Delaunay_triangulation). We need to convert the set of points into a mesh which approximates the same surface that the points approximate, and one of the best was to do this is to make triangles out of any points that are close enough, while avoiding self-intersections and other non-manifold geometry. Thanks to these extra restrictions, there are a lot of edge cases you have to consider when making a mesh out of a scattering of points over a surface. 

Hence what @patowen suggested above, and what I elaborated on, where we take a spherical mesh (a surface guaranteed to be manifold) and shift vertices in ways that we can guarantee will avoid self-intersection, and subdivide.

So basically, I think it will be hard (or at least annoying) to generate a mesh from what your code produces; so if you want to try making your point-smoothing system produce meshes, you should figure out how to put edges between your vertices right when you generate them, and thus essentially be writing an implementation of the subdivide/morph idea discussed above.

**Comment by patowen** [comment]
We should also be careful to make sure the player has enough control over the shape and overall look of the microbe. One way to do this is to make sure that the mesh can tightly conform to hexes created by the player, but additional constraints would also be useful. Are we clear of our end goal, or does that need to be discussed in the forums?

**Comment by DrChewbacca** [comment]
How I have been doing it is that for every point it's angle(s) never change(s), only it's radial distance from the origin.

So any which are neighbours at the start are neighbours for all time and then can never overlap or cross. The manifold is diffeomorphic to a sphere at all times.

This does slightly reduce the number of shapes you can have but I think it makes things nice and simple.

"where we take a spherical mesh (a surface guaranteed to be manifold) and shift vertices in ways that we can guarantee will avoid self-intersection, and subdivide." - this is what I've been doing.

**Comment by Moopli** [comment]
All the people who need to be involved are here, no worries. What scheme do you have in mind for constraining the mesh to the player-created hexes?

**Comment by patowen** [comment]
Sorry for the lack of response, as I am not entirely sure of my scheme, yet. My previous comment was me voicing my concern that the player will be forced to play as an indistinct blob if we are not careful. I will experiment when I have time, but it involves meddling in an area of 3D graphics that I currently lack experience in.

**Comment by Moopli** [comment]
@DrChewbacca 
> this is what I've been doing.

Not quite -- a mesh has not only the vertices but also the triangles connecting them. You have a vertex set which you subdivide and smooth, not a mesh.

@patowen I see what you mean -- a membrane defined by, say, NURBS, would be very smooth and boring unless we either add more control points or use some sort of trickery. Remember though: our goal for membrane rendering is to use a shader to make the membrane nearly transparent except when viewed close to tangentially, so the membrane looks something like what one would see in an optical microscope or TEM. We can probably avoid worrying almost entirely about the near and far surfaces, and just make sure they don't do anyhting *too* unreasonable.

**Comment by patowen** [comment]
That's true. Giving the user control of the microbe's look in the third dimension, which the user will never see, isn't really necessary. I'm pretty sure the main problem I currently have with NURBS is that I'm not sure how easy that is to implement and how possible it will be to correctly smooth it out. A lot of this probably just comes from lack of experience.

**Comment by Moopli** [comment]
NURBS aren't easy to mesh/render at all. You basically have to build a recursive subdivision surface to create more and more vertices that fit the NURBS curves, until the output mesh looks "good enough". This is why I've been trying to stick to a system that goes thorugh creating an implicit surface and then meshing that -- you can easily guarantee that it will be smooth and manifold, and you can use a bunch of fudge factors to make the implicit function cheap to calculate. 

Of course, I'm likely just more in my comfort zone when working with implicit surfaces.

**Comment by patowen** [comment]
That's true. The problem I find with implicit functions is that we also have to decide which points to plug in and do something like marching cubes. If we want a fine mesh and we do it naively, we'll have a dense 3D grid of points, which will likely be too slow. I used marching cubes with the metaballs in my original implementation, and that lagged a bit when the mesh was being modified.

There's likely no one best way of doing this. Perhaps we should lightly consider the creature stage in case we want some similarities between the membrane and skin calculations. I highly doubt hexes will be used in anything but the 2D stages.

**Comment by Moopli** [comment]
Well, to put it simply, then we don't do it naively, and don't generate a fine mesh.

Given that we require the membrane to be simply-connected, then we can generate the mesh by projecting the vertices of a simpler mesh (say, a sphere centered on the origin) outwards, either along the gradient, or along straight lines away from the origin. Then we apply some simple subdivision criterion (for example, by inserting new vertices in the middle of long edges), and repeat the projection and subdivision on the new vertices a few times.

**Comment by Moopli** [comment]
And now for a worked example: 

Consider a membrane which the user controls using the method outlined above:
![Rabies' Concept](https://cloud.githubusercontent.com/assets/7467906/2967641/ee91ed30-db32-11e3-9b24-448562c1f373.png)
Suppose the curve is kept bilaterally symmetric, and the user cannot make a curve that "doubles back on itself" from the point of view of the nucleus (which we can justify biologically somewhat, as a cytoskeleton has all microtubules anchored right by the nucleus, they don't bend much, and they're the main determinors of the large-scale shape of the cell -- yes, we ignore neurons, but that's okay as they don't swim freely). Then, to give us our third dimension, we rotate this curve around the axis of the cell.

This entire process is simple enough that we need not actually ever build a volume to then produce a mesh from -- we can just evaluate, in-place, the target position of any particular vertex:

    interpolate(x,y,z) ->
        # where y is along axis of symmetry, x and z are interchangeable due to symmetry
        theta = atan(hypot(x,z), y)
        newdistance = control_curve.get(theta)
        scale = newdistance / hypot(x,y,z)
        x,y,z *= scale
        return x,y,z

**Comment by patowen** [comment]
I like the idea of not building a volume and just building the mesh outright, and I do know how to rotate curves around an axis. Unfortunately, I do not think that rotating curves around an axis gives players the effect they would want, since ridges would be common. I also like the idea of editing the outline of the cell rather than having the membrane determined by the hexes. I'm not sure why I was against the idea earlier. Maybe I overestimated how good metaballs were.

I'll try to think of some other way of introducing the third dimension in a way that is neither NURBS nor revolving around an axis (which would also require symmetry). I doubt think such an algorithm will be trivial, but coming up with something good should be doable. It would probably start with something like NURBS, but smoother to begin with (with fudge factors).

**Comment by DrChewbacca** [comment]
I'm not really sure that I'm helping. I had another go with filling in the triangles, results are here.

http://thrivegame.forum-free.ca/t1508-cell-membrane#31533

Feel free to ignore me if I'm way out of my depth and not understanding the problem properly. Thanks for your time.
